benchmark: imply flags with comments in benchmarks

joyeecheung opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

benchmark

##### Description of change
<!-- Provide a description of the change below this comment. -->

This PR makes it possible for the benchmarks to indicate what execution flags should be passed in when they are run, similar to the way tests do this.

`benchmark/common.js` makes it possible to pass flags by making benchmarks run "inherit" the flags passed to `benchmark/compare.js`, but IMHO this information should be included inside the benchmark code, and it makes it easier to automate the running of benchmarks, if we want to do something like https://github.com/nodejs/node/issues/8157.

mscdex commented:
IMHO just explicitly adding `--expose-internals` like common.js does is enough.

joyeecheung commented:
@mscdex There might be other flags needed in the future(e.g. from what I've grepped from the tests, GC flags, harmony flags, warning flags, `--zero-fill-buffers`, `--max_old_space_size`)

Trott commented:
/cc @nodejs/benchmarking @AndreasMadsen 

(Aside: Do we need to update @nodejs/benchmarking membership? Kind of surprised @AndreasMadsen isn't in there, for starters.)

addaleax commented:
> (Aside: Do we need to update @nodejs/benchmarking membership? Kind of surprised @AndreasMadsen isn't in there, for starters.)

Pinging @nodejs/benchmarking is pretty weird anyway, even @mscdex isn’t on it.

AndreasMadsen commented:
It is much better to implement these things in `common.js` (see: [benchmark/common.js#L141](https://github.com/nodejs/node/blob/master/benchmark/common.js#L141)). The reason is that we allow many different entry points to the benchmarks:

```
node benchmark/scatter.js
node benchmark/compare.js
node benchmark/util/normalize-encoding.js
```

so just adding this to `compare.js` doesn't fully solve the problem, but `common.js` is used in all cases.

If I understand correctly the problem is to call `require('internal/')`, the solution is to just put the require inside the `main` function.

AndreasMadsen commented:
@Trott the benchmarking working group has very little to do with `benchmark/`, they run their own benchmark suite which focuses on monitoring performance between node releases. That is why I'm not in the WG. The seperation is a little inconvenient because it makes it very difficult to get momentum behind: https://github.com/nodejs/benchmarking/pull/58

joyeecheung commented:
@AndreasMadsen I looked into the code a little bit futher, and If I understand correctly, `compare.js`/`scatter.js`/`run.js` pass in the flags when the benchmarks are compiled, while `common.js` pass in the flags when the benchmarks are run. That means if we want to get the code compiled under a certain flag, we have to either put the code inside `main`, or pass in the flags to `compare.js`/`scatter.js`/`run.js`. Simply moving the logic to `common.js` still doesn't make it possible to control the flags in compile time.

Maybe the hints can be provided as:

```js
// This gets passed in both phases
// Flags: --expose-internals 

// This gets passed when main is run, implemented in common.js
// RuntimeFlags: --max_old_space_size=1024

// This gets passed when main is compiled, implemented in compare.js and friends
// CompileTimeFlags: --harmony
```

This also makes it possible to pass other flags than the hard-corded `--expose-internals` at runtime, should we want to use them in the future(e.g. `--allow_natives_syntax`, although it can be passed via `v8.setFlagsFromString`, but not all the flags are v8 flags or can be suddenly turned on without major impact to the program).

joyeecheung commented:
@AndreasMadsen I looked into the code a little bit futher, and If I understand correctly, `compare.js`/`scatter.js`/`run.js` pass in the flags when the benchmarks are compiled, while `common.js` pass in the flags when the benchmarks are run. That means if we want to get the code compiled under a certain flag, we have to either put the code inside `main`, or pass in the flags to `compare.js`/`scatter.js`/`run.js`. Simply moving the logic to `common.js` still doesn't make it possible to control the flags in compile time.

Maybe the hints can be provided as:

```js
// This gets passed in both phases
// Flags: --expose-internals 

// This gets passed when main is run, implemented in common.js
// RuntimeFlags: --max_old_space_size=1024

// This gets passed when main is compiled, implemented in compare.js and friends
// CompileTimeFlags: --harmony
```

This also makes it possible to pass other flags than the hard-corded `--expose-internals` at runtime, should we want to use them in the future(e.g. `--allow_natives_syntax`, although it can be passed via `v8.setFlagsFromString`, but not all the flags are v8 flags or can be suddenly turned on without major impact to the program).

If you think this unnecessary, please let me know. But even if we don't implement this feature, I think at least the benchmark doc should mention this caveat (`require('internal/*')` inside `main`).

AndreasMadsen commented:
@joyeecheung I'm not sure what the difference between compiling and running is in a JIT context.

My guess is that you think `--expose_internals` is somehow added during runtime, this is not the case. A benchmark script actually forks itself, which happens in [benchmark/common.js#L141](https://github.com/nodejs/node/blob/master/benchmark/common.js#L141). This is to ensure that each benchmark configuration gets a fresh v8 instance.

Whatever solution we come up with it should be possible to run benchmarks using just `node benchmark/util/normalize-encoding.js`. Otherwise it becomes complicated to profile them.

joyeecheung commented:
@AndreasMadsen By running and compiling I was referring to the running/compiling of the benchmark script. So I looked into the code and turns out I somehow had a delusion that the script would get compiled once in `https://github.com/nodejs/node/blob/master/benchmark/common.js#L26` and run once in `https://github.com/nodejs/node/blob/master/benchmark/common.js#L24`, but in fact it's compiled/run twice. So all the code in the benchmark script compiled by V8 in the first pass would not be relevant to those in the second pass. My bad :P. This is not a problem if the code outside `main` doesn't have side effects or requires special flags to run(like `require`ing a internal module), but IMO the doc should mention that any code beyond declarations should be put inside `main`.

> Whatever solution we come up with it should be possible to run benchmarks using just node benchmark/util/normalize-encoding.js. Otherwise it becomes complicated to profile them.

I agree. So if we still keep the `// Flags:` stuff, it should be implemented in `common.js`, as you said earlier. And the docs can be updated to demonstrate what kinds of code goes where.

AndreasMadsen commented:
> This is not a problem if the code outside main doesn't have side effects or requires special flags to run(like requireing a internal module), but IMO the doc should mention that any code beyond declarations should be put inside main.

Sounds like a good idea, I think [benchmark/README.md#creating-a-benchmark](https://github.com/nodejs/node/blob/master/benchmark/README.md#creating-a-benchmark) would be the ideal place.

joyeecheung commented:
@AndreasMadsen  So I revisited this PR again today and realized that if we want to make sure:

> Whatever solution we come up with it should be possible to run benchmarks using just node benchmark/util/normalize-encoding.js. Otherwise it becomes complicated to profile them.

Then the best way for benchmarks to specify the flags that they wish to be run with is to pass them into `createBenchmark`. I've updated this PR to this direction. Let me know what you think :)

AndreasMadsen commented on benchmark/common.js:
```diff
@@ -138,7 +139,7 @@ Benchmark.prototype._run = function() {
     const child = child_process.fork(require.main.filename, childArgs, {
       env: childEnv,
-      execArgv: ['--expose_internals'].concat(process.execArgv)
+      execArgv: ['--expose_internals', ...self.flags, ...process.execArgv]
```
`common.js` needs to run on old node versions.

joyeecheung replied:
Thanks for this heads up :)

AndreasMadsen commented on benchmark/common.js:
```diff
@@ -3,17 +3,18 @@
 
-exports.createBenchmark = function(fn, options) {
-  return new Benchmark(fn, options);
+exports.createBenchmark = function(fn, options, flags) {
```
I think it would be better to make it `exports.createBenchmark = function(fn, config, options)`, where `options` is an object e.g. `{ flags: [] }`. I don't know why `config` is called `options` here, when it is not really optional.

joyeecheung replied:
I agree. I also noticed that the docs name the second argument `configs`, so I will use `configs` as the local variable name here.

AndreasMadsen commented on benchmark/README.md:
```diff
@@ -299,18 +299,44 @@ the runtime of the benchmark. Run `bench.start()` after the initialization
+run the benchmarks with will be used. In the second pass,
+the `main` function will be run, and the process will be run with:
+
+*  `--expose_internals`
```
I think we should simplify it, such that `--expose_internals` is added explicitly in whatever tests it is required. You can just search for `require('/internal`.

joyeecheung replied:
Yeah, I think it's simpler. (And from what I've grepped, there is only one benchmark uses this :P)

joyeecheung commented:
@AndreasMadsen I've updated the commits, with some more comments added to the properties of `Benchmark` objects. PTAL and see if there is any misunderstanding about these properties.

AndreasMadsen commented on benchmark/README.md:
```diff
@@ -287,37 +287,88 @@ chunk     encoding       mean confidence.interval
 const SlowBuffer = require('buffer').SlowBuffer;
 
-const bench = common.createBenchmark(main, {
+// If you want to benchmark the internal modules, add
```
nit: you ready wrote this in the `options` object and in the `main` function. I think saying it once is fine.

joyeecheung replied:
Addressed :)

joyeecheung commented:
Ping @mscdex, I see @AndreasMadsen requested a review.

mscdex commented:
Linter: https://ci.nodejs.org/job/node-test-linter/6134/

mscdex commented:
LGTM

joyeecheung commented:
If there are no more concerns about this I will going to land it. Inclined to land as two separate commits because the commit message could be a tad long if concatenated.

gibfahn commented:
Two approvals and no issues, landing sounds good to me.

I'd just go for one commit, I don't think a long commit log is an issue (if you feel it's going into too much detail you can always remove some of it).

MylesBorins commented:
This does not land cleanly in LTS. Added dont-land label. Please feel free to manually backport

AndreasMadsen commented:
It should not be backported, the entire benchmark suite was rewritten in v7 and declared a major change.
